packages/protocol/contracts/verifiers/SgxVerifier.sol

```diff
@@ -143,30 +143,43 @@ contract SgxVerifier is EssentialContract, IVerifier {
     }
 
     /// @inheritdoc IVerifier
-    function verifyProof(
-        Context calldata _ctx,
-        TaikoData.Transition calldata _tran,
+    function verifyProofs(
+        Context[] calldata _ctx,
         TaikoData.TierProof calldata _proof
     )
         external
         onlyFromNamed(LibStrings.B_TAIKO)
     {
-        // Do not run proof verification to contest an existing proof
-        if (_ctx.isContesting) return;
-
         // Size is: 89 bytes
-        // 4 bytes + 20 bytes + 65 bytes (signature) = 89
-        if (_proof.data.length != 89) revert SGX_INVALID_PROOF();
+        // 4 bytes + 20 bytes + 20 bytes + 65 bytes (signature) = 109
+        if (_proof.data.length != 109) revert SGX_INVALID_PROOF();
 
         uint32 id = uint32(bytes4(_proof.data[:4]));
-        address newInstance = address(bytes20(_proof.data[4:24]));
-
-        address oldInstance = ECDSA.recover(
-            LibPublicInput.hashPublicInputs(
-                _tran, address(this), newInstance, _ctx.prover, _ctx.metaHash, taikoChainId()
-            ),
-            _proof.data[24:]
-        );
+        address oldInstance = address(bytes20(_proof.data[4:24]));
+        address newInstance = address(bytes20(_proof.data[24:44]));
+        bytes memory signature = _proof.data[44:];
+
+        // Collect public inputs
+        bytes32[] memory public_inputs = new bytes32[](_ctx.length + 1);
+        // First public input is the current instance public key
+        public_inputs[0] = bytes32(bytes20(oldInstance));
```

Comment:
Is this necessary? I mean as oldInstance is always recovered from signature, why we need explicit list oldInstance here?

Reply:
We need to have it available in the aggregation proof, because that's the address we need to verify against for the first block proof. We need to get it from somewhere to be able to correctly verify that proof, because we don't have it available anymore in the SGX instance itself (could be many blocks ago, and we only store the latest key pair in SGX).